File: src/payment_fee_service/providers/stripe/provider.py
from __future__ import annotations

from decimal import Decimal, InvalidOperation
from typing import Any

from payment_fee_service.data.snapshots import ProviderSnapshot
from payment_fee_service.domain.errors import (
    AmbiguousRulesError,
    InsufficientContextError,
    QuoteUnavailableError,
    UnknownMarketError,
)
from payment_fee_service.domain.models import (
    CapabilityInfo,
    MarketInfo,
    QuoteRequest,
    StripeQuoteRequest,
)
from payment_fee_service.domain.rules import CompiledFeePlan, ExecutableFeeRule
from payment_fee_service.providers.stripe.matcher import (
    build_context,
    financial_signature,
    missing_dimensions,
    rule_matches_known_context,
    specificity,
)

NON_TRANSACTION_FIELDS = {"payout_type", "dispute_state"}


class StripeProvider:
    provider_id = "stripe"

    def __init__(self, snapshot: ProviderSnapshot) -> None:
        self.snapshot = snapshot
        self._markets = {
            str(item.get("account_country", "")).upper(): item
            for item in snapshot.core.get("markets", [])
        }
        self._index = {
            str(item.get("account_country", "")).upper(): item
            for item in snapshot.index.get("markets", [])
        }
        self._payment_methods = self._extract_payment_methods(snapshot.payment_methods or {})

    def compile_rules(self, request: QuoteRequest) -> CompiledFeePlan:
        if not isinstance(request, StripeQuoteRequest):
            raise TypeError("StripeProvider received a non-Stripe request")
        market = self._markets.get(request.account_country)
        if market is None:
            raise UnknownMarketError(self.provider_id, request.account_country)
        context = build_context(request)
        all_rules = [rule for rule in market.get("rules", []) if self._is_transaction_rule(rule)]

        relevant = [
            rule
            for rule in all_rules
            if self._method_scope_matches(rule, request.payment.method)
            and self._classification_usable(rule)
            and self._has_fee_value(rule)
        ]
        potentially_matching = [
            rule for rule in relevant if self._matches_non_missing(rule, context)
        ]
        missing = sorted(
            {
                dimension
                for rule in potentially_matching
                for dimension in missing_dimensions(rule, context)
            }
        )
        if missing:
            raise InsufficientContextError(
                [self._api_field_name(item) for item in missing],
                provider="stripe",
                candidate_rule_ids=[rule.get("rule_id") for rule in potentially_matching],
            )

        matched = [rule for rule in relevant if rule_matches_known_context(rule, context)]
        base = [
            rule
            for rule in matched
            if self._same_method(rule.get("payment_method"), request.payment.method)
        ]
        additives = [
            rule
            for rule in matched
            if rule.get("payment_method") is None and self._is_contextual_additive(rule)
        ]
        selected = self._resolve_base(base) + self._deduplicate(additives)
        if not selected:
            raise QuoteUnavailableError(
                "No classified Stripe transaction fee rule matched the supplied context.",
                provider="stripe",
                market=request.account_country,
                payment_method=request.payment.method,
            )

        index = self._index.get(request.account_country, {})
        rules = [self._to_executable(rule, request.amount.currency) for rule in selected]
        assumptions = [
            "Public standard pricing was used; negotiated or IC++ pricing is not represented.",
            "The published dataset does not encode provider settlement rounding, so "
            "standard currency rounding is used.",
        ]
        return CompiledFeePlan(
            provider=self.provider_id,
            market=request.account_country,
            currency=request.amount.currency,
            rules=rules,
            assumptions=assumptions,
            schema_version=self.snapshot.schema_version,
            content_sha256=index.get("content_sha256"),
            source_urls=list(index.get("source_urls") or []),
            source_updated_at=index.get("source_updated_at"),
            data_ref=self.snapshot.data_ref,
        )

    def _resolve_base(self, rules: list[dict[str, Any]]) -> list[dict[str, Any]]:
        if not rules:
            return []
        max_specificity = max(specificity(rule) for rule in rules)
        most_specific = [rule for rule in rules if specificity(rule) == max_specificity]
        signatures = {financial_signature(rule) for rule in most_specific}
        if len(signatures) > 1:
            raise AmbiguousRulesError([str(rule.get("rule_id")) for rule in most_specific])
        return [sorted(most_specific, key=lambda rule: str(rule.get("rule_id")))[0]]

    @staticmethod
    def _deduplicate(rules: list[dict[str, Any]]) -> list[dict[str, Any]]:
        unique: dict[tuple[Any, ...], dict[str, Any]] = {}
        for rule in sorted(rules, key=lambda item: str(item.get("rule_id"))):
            key = (
                financial_signature(rule),
                rule.get("name"),
                tuple(
                    sorted(
                        (c.get("dimension"), str(c.get("value")))
                        for c in rule.get("conditions", [])
                    )
                ),
            )
            unique.setdefault(key, rule)
        return list(unique.values())

    @staticmethod
    def _is_transaction_rule(rule: dict[str, Any]) -> bool:
        if any(rule.get(field) is not None for field in NON_TRANSACTION_FIELDS):
            return False
        return str(rule.get("unit", "per_transaction")) == "per_transaction"

    @staticmethod
    def _classification_usable(rule: dict[str, Any]) -> bool:
        status = str(rule.get("classification_status", "unclassified")).lower()
        return status not in {"", "unknown", "unclassified"}

    @staticmethod
    def _has_fee_value(rule: dict[str, Any]) -> bool:
        return any(
            rule.get(field) is not None
            for field in (
                "basis_points",
                "percentage",
                "fixed_amount",
                "minimum_amount",
                "maximum_amount",
            )
        )

    @staticmethod
    def _method_scope_matches(rule: dict[str, Any], requested_method: str) -> bool:
        method = rule.get("payment_method")
        if method is not None:
            return isinstance(method, str) and method.casefold() == requested_method.casefold()
        card_scoped = any(
            rule.get(field) is not None for field in ("card_origin", "card_region", "card_tier")
        )
        return not card_scoped or requested_method.casefold() == "card"

    @staticmethod
    def _same_method(value: Any, requested: str) -> bool:
        return isinstance(value, str) and value.casefold() == requested.casefold()

    @staticmethod
    def _matches_non_missing(rule: dict[str, Any], context: dict[str, Any]) -> bool:
        # Temporarily fill missing constrained values from the rule itself. This identifies rules
        # that could match once the caller supplies the missing context, without accepting them.
        probe = dict(context)
        for field, value in rule.items():
            if value is not None and probe.get(field) is None:
                probe[field] = value
        for condition in rule.get("conditions", []):
            dimension = condition.get("dimension")
            if dimension and probe.get(dimension) is None:
                probe[dimension] = condition.get("value")
        try:
            return rule_matches_known_context(rule, probe)
        except QuoteUnavailableError:
            raise

    @staticmethod
    def _is_contextual_additive(rule: dict[str, Any]) -> bool:
        return bool(
            rule.get("conditions")
            or rule.get("card_origin") is not None
            or rule.get("card_region") is not None
            or rule.get("card_tier") is not None
            or rule.get("currency_conversion_required") is not None
            or rule.get("customer_country") is not None
        )

    @staticmethod
    def _to_executable(rule: dict[str, Any], currency: str) -> ExecutableFeeRule:
        def decimal_or_none(field: str) -> Decimal | None:
            value = rule.get(field)
            if value is None:
                return None
            try:
                return Decimal(str(value))
            except InvalidOperation as exc:
                raise QuoteUnavailableError(
                    "A selected Stripe fee rule contains an invalid decimal.",
                    rule_id=rule.get("rule_id"),
                    field=field,
                    value=value,
                ) from exc

        behavior = str(rule.get("behavior", "additive"))
        if behavior not in {"additive", "base", "standard"}:
            raise QuoteUnavailableError(
                "Unsupported Stripe fee-rule behavior.",
                rule_id=rule.get("rule_id"),
                behavior=behavior,
            )
        fixed_currency = rule.get("fixed_currency")
        return ExecutableFeeRule(
            rule_id=str(rule["rule_id"]),
            label=str(rule.get("name") or rule.get("source_text") or "Stripe processing fee"),
            percentage=decimal_or_none("percentage"),
            basis_points=decimal_or_none("basis_points"),
            fixed_amount=decimal_or_none("fixed_amount"),
            fixed_currency=str(fixed_currency).upper() if fixed_currency else currency,
            minimum_amount=decimal_or_none("minimum_amount"),
            maximum_amount=decimal_or_none("maximum_amount"),
            behavior=behavior,
            classification_status=str(rule.get("classification_status", "classified")),
            confidence=float(rule["confidence"]) if rule.get("confidence") is not None else None,
            exactness=rule.get("exactness"),
            source_url=rule.get("source_url"),
        )

    def markets(self) -> list[MarketInfo]:
        return [
            MarketInfo(
                provider=self.provider_id,
                account_country=country,
                market_code=str(market.get("stripe_market_code", country.lower())),
                locale=market.get("locale"),
                status=str(market.get("derivation_status", "unclassified")),
                source_urls=list(self._index.get(country, {}).get("source_urls") or []),
            )
            for country, market in sorted(self._markets.items())
        ]

    def capabilities(self, account_country: str) -> CapabilityInfo:
        country = account_country.upper()
        market = self._markets.get(country)
        if market is None:
            raise UnknownMarketError(self.provider_id, country)
        methods = sorted(
            {
                str(rule["payment_method"])
                for rule in market.get("rules", [])
                if rule.get("payment_method")
                and self._classification_usable(rule)
                and self._has_fee_value(rule)
            }
        )
        required = sorted(
            {
                self._api_field_name(field)
                for rule in market.get("rules", [])
                for field in (
                    "card_origin",
                    "card_region",
                    "card_tier",
                    "channel",
                    "recurring",
                    "billing_type",
                    "currency_conversion_required",
                )
                if rule.get(field) is not None
            }
        )
        return CapabilityInfo(
            provider=self.provider_id,
            account_country=country,
            quotable=bool(methods),
            payment_methods=methods or sorted(self._payment_methods.get(country, set())),
            required_context=required,
        )

    @staticmethod
    def _api_field_name(dimension: str) -> str:
        mapping = {
            "payment_method": "payment.method",
            "card_origin": "payment.card.origin",
            "card_region": "payment.card.region",
            "card_tier": "payment.card.tier",
            "channel": "payment.channel",
            "recurring": "payment.recurring",
            "billing_type": "payment.billing_type",
            "currency_conversion_required": "payment.currency_conversion_required",
            "presentment_currency": "amount.currency",
            "settlement_currency": "settlement_currency",
            "customer_country": "customer_country",
        }
        return mapping.get(dimension, f"payment.context.{dimension}")

    @staticmethod
    def _extract_payment_methods(document: dict[str, Any]) -> dict[str, set[str]]:
        result: dict[str, set[str]] = {}
        methods = document.get("payment_methods") or document.get("methods") or []
        for method in methods:
            method_id = method.get("method_id") or method.get("id")
            if not method_id:
                continue
            countries = (
                method.get("account_countries") or method.get("supported_account_countries") or []
            )
            for country in countries:
                result.setdefault(str(country).upper(), set()).add(str(method_id))
        return result

File: packages/payment-fee-typescript/src/providers/stripe/provider.ts
import { Decimal } from "decimal.js";
import { InsufficientTransactionContext, QuoteNotAvailable, UnknownMarket } from "../../errors.js";
import type { ExecutableRule } from "../../calculator.js";
import type { QuoteRequest, StripeQuoteRequest } from "../../models.js";
import type { StripeRule } from "./condition-matcher.js";
import {
  apiFieldName,
  conditionStatus,
  executableFromRule,
  isEvaluable,
  normalizeConditions,
  selectAdditiveRules,
  sortBySpecificityDesc,
} from "./condition-matcher.js";

export interface StripeMarket {
  account_country: string;
  rules?: StripeRule[];
}

export interface StripeCore {
  schema_version?: number;
  markets?: StripeMarket[];
}

export interface StripeIndexMarket {
  account_country: string;
  content_sha256?: string | null;
  source_urls?: string[];
  source_updated_at?: string | null;
}

export interface StripeIndex {
  schema_version?: number;
  markets?: StripeIndexMarket[];
}

export class StripeProvider {
  private readonly markets = new Map<string, StripeMarket>();

  constructor(private readonly core: StripeCore) {
    for (const market of core.markets ?? []) {
      this.markets.set(market.account_country.toUpperCase(), market);
    }
  }

  compileRules(request: QuoteRequest): ExecutableRule[] {
    const stripeRequest = request as StripeQuoteRequest;
    const market = this.markets.get(stripeRequest.account_country.toUpperCase());
    if (!market) {
      throw new UnknownMarket("stripe", stripeRequest.account_country);
    }

    const currency = stripeRequest.amount.currency;
    const context = buildContext(stripeRequest);

    const fullMatches: { rule: StripeRule; specificity: number }[] = [];
    const missingMatches: { rule: StripeRule; missing: string[]; specificity: number }[] = [];

    for (const rule of market.rules ?? []) {
      const conditions = normalizeConditions(rule);
      let conflict = false;
      const missing: string[] = [];
      for (const condition of conditions) {
        const status = conditionStatus(condition, context);
        if (status === "conflict") {
          conflict = true;
          break;
        }
        if (status === "missing") {
          missing.push(apiFieldName(condition.dimension));
        }
      }
      if (conflict) continue;
      const specificity = conditions.length;
      if (missing.length > 0) {
        missingMatches.push({ rule, missing: [...new Set(missing)].sort(), specificity });
      } else {
        fullMatches.push({ rule, specificity });
      }
    }

    if (fullMatches.length === 0) {
      if (missingMatches.length > 0) {
        const allMissing = [...new Set(missingMatches.flatMap((m) => m.missing))].sort();
        throw new InsufficientTransactionContext(allMissing, {
          provider: "stripe",
          market: stripeRequest.account_country,
        });
      }
      throw new QuoteNotAvailable("No Stripe fee rule matched the supplied context.", {
        provider: "stripe",
        market: stripeRequest.account_country,
      });
    }

    const maxSpec = Math.max(...fullMatches.map((m) => m.specificity));
    const mostSpecific = fullMatches.filter((m) => m.specificity === maxSpec).map((m) => m.rule);

    if (!mostSpecific.some((r) => isEvaluable(r))) {
      throw new QuoteNotAvailable("The most specific matching Stripe fee rule cannot be quoted.", {
        provider: "stripe",
        market: stripeRequest.account_country,
        rule_ids: mostSpecific.map((r) => r.rule_id),
      });
    }

    const baseCandidates = sortBySpecificityDesc(fullMatches)
      .map((m) => m.rule)
      .filter((r) => isEvaluable(r) && (r.behavior ?? "base") !== "additive");

    if (baseCandidates.length === 0) {
      throw new QuoteNotAvailable("No evaluable base Stripe fee rule matched.", {
        provider: "stripe",
        market: stripeRequest.account_country,
      });
    }

    const base = baseCandidates[0];
    const additiveRules = selectAdditiveRules(market.rules ?? [], context);

    const rules = [executableFromRule(base, currency)];
    for (const rule of additiveRules) {
      rules.push(executableFromRule(rule, currency));
    }

    return rules;
  }

  auditContract(): Record<string, number> {
    let total = 0;
    let parsed = 0;
    let skipped = 0;
    let contextRequired = 0;

    for (const market of this.core.markets ?? []) {
      for (const rule of market.rules ?? []) {
        total += 1;
        if (!isEvaluable(rule)) {
          skipped += 1;
          continue;
        }
        if ((rule.fee_components ?? []).length === 0 && !rule.basis_points && !rule.fixed_amount) {
          skipped += 1;
          continue;
        }
        if ((rule.conditions ?? []).length > 0) {
          contextRequired += 1;
        }
        parsed += 1;
      }
    }

    return {
      stripe_calculable_rules_total: total,
      stripe_calculable_rules_parsed: parsed,
      stripe_calculable_rules_skipped: skipped,
      stripe_context_required: contextRequired,
    };
  }
}

function isNumeric(value: unknown): boolean {
  return typeof value === "number" || (typeof value === "string" && !Number.isNaN(Number(value)));
}

function contextValuesEqual(a: unknown, b: unknown): boolean {
  if (typeof a === "boolean" || typeof b === "boolean") {
    return Boolean(a) === Boolean(b);
  }
  if (isNumeric(a) && isNumeric(b)) {
    try {
      return new Decimal(String(a)).eq(new Decimal(String(b)));
    } catch {
      return false;
    }
  }
  return a === b;
}

function buildContext(request: StripeQuoteRequest): Record<string, unknown> {
  const t = request.transaction;
  const context: Record<string, unknown> = {
    account_country: request.account_country.toUpperCase(),
    customer_country: request.customer_country?.toUpperCase() ?? null,
    amount_currency: request.amount.currency.toUpperCase(),
    transaction_amount: request.amount.value,
    presentment_currency: request.amount.currency.toUpperCase(),
    settlement_currency: request.settlement_currency?.toUpperCase() ?? null,
    product_id: t.product_id ?? null,
    variant_id: t.variant_id ?? null,
    payment_method: t.payment_method ?? null,
    payment_method_variant: t.payment_method_variant ?? null,
    channel: t.channel ?? null,
    pricing_plan: t.pricing_plan ?? null,
    pricing_tier: t.pricing_tier ?? null,
    payer: t.payer ?? null,
    unit: t.unit ?? "per_transaction",
    currency_conversion_required: t.currency_conversion_required ?? null,
    recurring: t.recurring ?? null,
    billing_type: t.billing_type ?? null,
    transaction_region: t.transaction_region ?? null,
    cross_border: t.cross_border ?? null,
    integration_type: t.integration_type ?? null,
    product_feature: t.product_feature ?? null,
    contract_length: t.contract_length ?? null,
    feature_enabled: t.feature_enabled ?? null,
    dispute_state: t.dispute_state ?? null,
    success: true,
  };

  if (t.card) {
    context.card_origin = t.card.origin ?? null;
    context.card_region = t.card.region ?? null;
    context.card_type = t.card.type ?? null;
    context.card_network = t.card.network ?? null;
    context.card_tier = t.card.tier ?? null;
    context.card_entry_mode = t.card.entry_mode ?? null;
  }

  if (t.settlement) {
    context.settlement_currency ??= t.settlement.currency?.toUpperCase();
    context.settlement_timing = t.settlement.timing ?? null;
  }

  if (t.bank) {
    context.bank_account_validation = t.bank.validation ?? null;
    context.bank_transfer_type = t.bank.transfer_type ?? null;
  }

  if (t.context?.success !== undefined) {
    context.success = t.context.success;
  }

  for (const [key, value] of Object.entries(t.context ?? {})) {
    if (value === undefined) {
      continue;
    }
    const existing = context[key];
    if (existing !== undefined && existing !== null) {
      if (!contextValuesEqual(value, existing)) {
        throw new QuoteNotAvailable("Contradictory duplicate value in transaction context.", {
          field: key,
          typed_value: existing,
          context_value: value,
        });
      }
    } else {
      context[key] = value;
    }
  }

  return context;
}
